generate/avro.go

```diff
@@ -0,0 +1,304 @@
+package generate
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"os"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"golang.org/x/tools/go/ast/astutil"
+)
+
+const (
+	avroHeader           = `{"name":%q,"type":"record","fields":[{"name":"AvroWriteTime","doc":"The timestamp when this avro data is written. Useful for identifying the newest row of data sharing keys.","type":"long","logicalType":"timestamp-millis"},{"name":"AvroDeleted","doc":"This is set to true when the Avro data is recording a delete in the source data.","default":false,"type":"boolean"},`
+	embeddedStructHeader = `{"name":%q,"type":"record","fields":[`
```

Comment:
I looked into using some of the code from a couple of Avro libraries but nothing I found was really built for this purpose. Additionally it seemed best not to tie us to a particular Avro library in step 1 of this process.

I then considered a few different ways of generating this schema but ultimately decided on parsing the Go struct for two reasons. First, I thought it would result in simpler more clear code to just parse that and print out the JSON for the Avro schema than it would to first build up an internal representation then write out the schema based on that. Now that it is done I'm not sure it is any clearer. The second was that I knew I was going to need to parse the go structs in step 3 of this process in order to generate helper functions for processing the code. I thought having some code that helps with parsing go structs already built would help then.

Reply:
Sounds good thanks for the explanation